### ISSUE TITLE
About text to semantic tokens

### ISSUE THREAD

**Issue by shanhaidexiamo** [collabora/WhisperSpeech#27]
Hi, thanks for your repo. I want to ask you about how to measure the effectiveness of the t2s model? Or to what extent does the loss converge to make the model usable?

Thanks

**Comment by jpc** [comment]
Hey, that's a very good question :)

One way I use to check if the T2S model works well is by decoding the semantic tokens back to text. You can see [an example of that in the inference notebook](https://colab.research.google.com/github/collabora/WhisperSpeech/blob/main/Inference%20example.ipynb#scrollTo=414d9b3c). It would be great to have a WER metric for that during training but we currently don't have it.

One way to think about it is that the S2A model is like a language model (generating speech) and the semantic tokens are just there to help it stay aligned with the input text (so it always chooses the correct completion instead of guessing the most probable one). They don't have to be perfect – it's enough if they are unambiguous.

PS. I ran an experiment lately where I trained a couple of T2S models recently that had substantially lower validation losses but using them did not seem to improve the quality of generated audio. I think there is a point where the model generates token sequences that are similar enough to the ground truth token streams that the S2A model is able to self-correct and you get diminishing returns. It is similar to how you train a language model with teacher forcing but then at some point it starts working well even with it's own predictions as context.

**Comment by shanhaidexiamo** [comment]
That's great! Thank you for clearing my mind. In fact, I did not perform the back-translation stage but directly fine-tune the pre-trained model because my ASR data is sufficient. Judging from your experimental conclusion, S2A model is the key to affecting the quality of generated audio,T2S model plays more of an alignment role. When I finish training the S2A model, I will combine it with the T2S to observe the quality of the generated audio. Thanks again for your reply.

**Comment by jpc** [comment]
Just to add that the T2S model also controls prosody in our approach. The semantic token extraction model also plays a large role and we've noticed that Whisper `medium` is a lot better than `base`.
